src/ucp/proto/proto_init.c

```diff
@@ -207,34 +247,18 @@ ucp_proto_init_add_memreg_time(const ucp_proto_common_init_params_t *params,
     if (context->rcache != NULL) {
         perf_factors[cpu_factor_id] =
                 ucs_linear_func_make(context->config.ext.rcache_overhead, 0);
+        perf_node = ucp_proto_perf_node_new_data("rcache lookup", "");
         return ucp_proto_perf_add_funcs(perf, range_start, range_end,
-                                        perf_factors, NULL, "rcache lookup",
-                                        "");
-    }
-
-    reg_perf_node = ucp_proto_perf_node_new_data("mem reg", "");
-
-    /* Go over all memory domains */
-    ucs_for_each_bit(md_index, reg_md_map) {
-        md_attr = &context->tl_mds[md_index].attr;
-        md_name = context->tl_mds[md_index].rsc.md_name;
-        ucs_linear_func_add_inplace(&perf_factors[cpu_factor_id],
-                                    md_attr->reg_cost);
-        ucs_trace("md %s reg: " UCP_PROTO_PERF_FUNC_FMT, md_name,
-                  UCP_PROTO_PERF_FUNC_ARG(&md_attr->reg_cost));
-        ucp_proto_perf_node_add_data(reg_perf_node, md_name, md_attr->reg_cost);
+                                        perf_factors, perf_node, NULL);
     }
 
-    if (!ucs_is_pow2(reg_md_map)) {
-        /* Multiple memory domains */
-        ucp_proto_perf_node_add_data(reg_perf_node, "total",
-                                     perf_factors[cpu_factor_id]);
-    }
-
-    status = ucp_proto_perf_add_funcs(perf, range_start, range_end,
-                                      perf_factors, reg_perf_node,
-                                      perf_node_name, "%u mds",
-                                      ucs_popcount(reg_md_map));
+    reg_perf_node = ucp_proto_init_memreg_node(context, reg_md_map,
+                                               &perf_factors[cpu_factor_id]);
+    perf_node     = ucp_proto_perf_node_new_data(perf_node_name, "%u mds",
+                                                 ucs_popcount(reg_md_map));
```

Comment:
shouldn't we count **non-zero** md map and not whole md map?
or maybe also show 0-cost MDs in the perf node for completeness